packages-exp/auth-exp/src/platform_cordova/popup_redirect/events.ts

```diff
@@ -41,6 +49,67 @@ export function _generateNewEvent(
   };
 }
 
+export function _savePartialEvent(auth: Auth, event: AuthEvent): Promise<void> {
+  return storage()._set(key(auth), (event as object) as PersistedBlob);
+}
+
+export async function _getAndRemoveEvent(
+  auth: Auth
+): Promise<AuthEvent | null> {
+  const event = (await storage()._get(key(auth))) as AuthEvent | null;
+  if (event) {
+    await storage()._remove(key(auth));
+  }
+  return event;
+}
+
+export function _eventFromPartialAndUrl(
+  partialEvent: AuthEvent,
+  url: string
+): AuthEvent | null {
+  // Parse the deep link within the dynamic link URL.
+  const callbackUrl = _getDeepLinkFromCallback(url);
+  // Confirm it is actually a callback URL.
+  // Currently the universal link will be of this format:
+  // https://<AUTH_DOMAIN>/__/auth/callback<OAUTH_RESPONSE>
+  // This is a fake URL but is not intended to take the user anywhere
+  // and just redirect to the app.
+  if (callbackUrl.includes('/__/auth/callback')) {
+    // Check if there is an error in the URL.
+    // This mechanism is also used to pass errors back to the app:
+    // https://<AUTH_DOMAIN>/__/auth/callback?firebaseError=<STRINGIFIED_ERROR>
+    const params = searchParamsOrEmpty(callbackUrl);
+    // Get the error object corresponding to the stringified error if found.
+    const errorObject = params['firebaseError']
+      ? JSON.parse(decodeURIComponent(params['firebaseError']))
```

Comment:
nit: is it a concern that `JSON.parse` may throw?

Reply:
Good point, done. I've added a helper that tries to parse or returns null